hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -3,321 +3,411 @@ package is.hail.expr.ir
-      if (i == 1) FastSeq(indexName -> TInt32) else empty
-    case AggFold(zero, _, _, accumName, otherAccumName, _) =>
-      if (i == 1) FastSeq(accumName -> zero.typ)
-      else if (i == 2) FastSeq(accumName -> zero.typ, otherAccumName -> zero.typ)
-      else empty
-    case NDArrayMap(nd, name, _) =>
-      if (i == 1) Array(name -> tcoerce[TNDArray](nd.typ).elementType) else empty
-    case NDArrayMap2(l, r, lName, rName, _, _) => if (i == 2) Array(
-        lName -> tcoerce[TNDArray](l.typ).elementType,
-        rName -> tcoerce[TNDArray](r.typ).elementType,
-      )
-      else empty
-    case CollectDistributedArray(contexts, globals, cname, gname, _, _, _, _) => if (i == 2)
-        Array(cname -> tcoerce[TStream](contexts.typ).elementType, gname -> globals.typ)
-      else empty
-    case TableAggregate(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case MatrixAggregate(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case TableFilter(child, _) => if (i == 1) child.typ.rowEnv.m else empty
-    case TableGen(contexts, globals, cname, gname, _, _, _) =>
-      if (i == 2) Array(cname -> elementType(contexts.typ), gname -> globals.typ)
-      else empty
-    case TableMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case TableMapRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
-    case TableAggregateByKey(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case TableKeyByAndAggregate(child, _, _, _, _) =>
-      if (i == 1) child.typ.globalEnv.m else if (i == 2) child.typ.rowEnv.m else empty
-    case TableMapPartitions(child, g, p, _, _, _) =>
-      if (i == 1) Array(g -> child.typ.globalType, p -> TStream(child.typ.rowType)) else empty
-    case MatrixMapRows(child, _) => if (i == 1) child.typ.rowEnv.bind("n_cols", TInt32).m else empty
-    case MatrixFilterRows(child, _) => if (i == 1) child.typ.rowEnv.m else empty
-    case MatrixMapCols(child, _, _) =>
-      if (i == 1) child.typ.colEnv.bind("n_rows", TInt64).m else empty
-    case MatrixFilterCols(child, _) => if (i == 1) child.typ.colEnv.m else empty
-    case MatrixMapEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
-    case MatrixFilterEntries(child, _) => if (i == 1) child.typ.entryEnv.m else empty
-    case MatrixMapGlobals(child, _) => if (i == 1) child.typ.globalEnv.m else empty
-    case MatrixAggregateColsByKey(child, _, _) =>
-      if (i == 1) child.typ.rowEnv.m else if (i == 2) child.typ.globalEnv.m else empty
-    case MatrixAggregateRowsByKey(child, _, _) =>
-      if (i == 1) child.typ.colEnv.m else if (i == 2) child.typ.globalEnv.m else empty
-    case BlockMatrixMap(_, eltName, _, _) => if (i == 1) Array(eltName -> TFloat64) else empty
-    case BlockMatrixMap2(_, _, lName, rName, _, _) =>
-      if (i == 2) Array(lName -> TFloat64, rName -> TFloat64) else empty
-    case _ => empty
-  }
+case class SegregatedBindingEnv[A, B](
```

Comment:
I guess what I'm asking is when is this useful?

Reply:
Also, there's this potential pitfall where:
```scala
env.mapValuesWithKey(f).unified != env.unified.mapValuesWithKey(f)
```

Reply:
It's because `Bindings` always adds new variables with their types, but not all consumers want that. For example, `ComputeUsesAndDefs` calls it like
```scala
          val bindings = Bindings.segregated(ir, i, env).mapNewBindings((_, _) => ir)
          if (!bindings.newBindings.allEmpty && !uses.contains(ir))
            uses.bind(ir, mutable.Set.empty[RefEquality[BaseRef]])
          compute(child, bindings.unified)
```
where `env: BindingEnv[IR]`. The call to `segregated` returns a `SegregatedBindingEnv[IR, Type]`, and then we replace the types with `ir` (the definition site of the new bindings).

It would be possible to avoid this by adding more implementations of `GenericBindingEnv` for each special use case like this, but I think this is simpler.

Reply:
> Also, there's this potential pitfall where:
> 
> ```scala
> env.mapValuesWithKey(f).unified != env.unified.mapValuesWithKey(f)
> ```

I don't follow. `mapValuesWithKey` isn't defined on `SegregatedBindingEnv`. If needed, we could promote it to `GenericBindingEnv`, in which case the implementation on `SegregatedBindingEnv` would need to map both parts. There is a `mapNewBindings` on `SegregatedBindingEnv`, but that's explicit that it only applies to the new bindings.

Reply:
What's the point of `(IR, Type)`? Consumers can just do `.typ` surely?

Reply:
> I don't follow. mapValuesWithKey isn't defined on SegregatedBindingEnv

My bad, I got confused with all these different envs.

Reply:
> What's the point of `(IR, Type)`? Consumers can just do `.typ` surely?

Sorry, where are you referring to?

Reply:
> The call to segregated returns a SegregatedBindingEnv[IR, Type], and then we replace the types with ir (the definition site of the new bindings).

Why not always use IR and then consumers can call `.typ` if they want the type?

Reply:
The IR that `ComputeUsesAndDefs` puts in the environment is the ir that defines the variable. The type of that node is not the type of the binding. In general there is no node whose type is the type of the binding. Consider `StreamMap(stream, "elt", body)`. `"elt"` is bound in body, with type the element type of `stream`, but `ComputeUsesAndDefs` wants the environment to map `"elt"` to the `StreamMap` node itself.